npm/webpack-dev-server-fresh/src/devServer.ts

```diff
@@ -0,0 +1,152 @@
+  return typeof error === 'string' ? error : error.message
+}
+
+/**
+ * import { devServer } from '@cypress/webpack-dev-server'
+ *
+ * Creates & returns a WebpackDevServer for serving files related
+ * to Cypress Component Testing
+ *
+ * @param config
+ */
+export function devServer (devServerConfig: WebpackDevServerConfig): Promise<Cypress.ResolvedDevServerConfig> {
+  return new Promise((resolve, reject) => {
+    const result = devServer.create(devServerConfig) as DevServerCreateResult
+
+    // When compiling in run mode
+    // Stop the clock early, no need to run all the tests on a failed build
+    result.compiler.hooks.done.tap('cyCustomErrorBuild', function (stats) {
+      if (stats.hasErrors()) {
+        const errors = stats.compilation.errors
+
+        devServerConfig.devServerEvents.emit('dev-server:compile:error', normalizeError(errors[0]))
+        if (devServerConfig.cypressConfig.isTextTerminal) {
+          process.exit(1)
+        }
+      }
+    })
+
+    if (result.version === 3) {
+      const srv = result.server.listen(0, '127.0.0.1', () => {
+        const port = (srv.address() as AddressInfo).port
+
+        debug('Component testing webpack server 3 started on port %s', port)
+
+        resolve({
+          port,
+          close: (done) => {
+            srv.close((err) => {
+              if (err) {
+                debug('closing dev server, with error', err)
+              }
+
+              debug('closed dev server')
+              done?.(err)
+            })
+          },
+        })
+      })
+
+      return
+    }
+
+    result.server.start().then(() => {
+      if (!result.server.options.port) {
+        return reject(new Error(`Expected port ${result.server.options.port} to be a number`))
+      }
+
+      debug('Component testing webpack server 4 started on port %s', result.server.options.port)
+
+      resolve({
+        port: result.server.options.port as number,
+        close: (done) => {
+          debug('closing dev server')
+          result.server.stop().then(() => done?.()).catch(done).finally(() => {
+            debug('closed dev server')
+          })
+        },
+      })
+    }).catch(reject)
+  })
+}
+
+/**
+ * Syncronously create the webpack server instance, without starting.
+ * Useful for testing
+ *
+ * @internal
+ */
+devServer.create = function (devServerConfig: WebpackDevServerConfig) {
+  const sourceWebpackModulesResult = sourceRelativeWebpackModules(devServerConfig)
+
+  let frameworkConfig: object | undefined
+
+  // If we have a framework specified, source the associated config
+  if (typeof devServerConfig.framework === 'string') {
+    switch (devServerConfig.framework) {
+      case 'create-react-app':
+        // frameworkConfig = createReactApp()
+        break
+      case 'react':
+        break
+      case 'nuxt':
+        break
+      default:
+        throw new Error(`Unexpected framework ${devServerConfig.framework}, expected one of ${ALL_FRAMEWORKS.join(', ')}`)
```

Comment:
Following my above comment - does this module really belong in `npm`? Or at least this part of it (the error handling, expecting some specific string arguments) - it doesn't seem like something we would not publish on npm, it's really opinionated and fairly useless outside of the binary.

Or is the separation more like:

- object API - internal only
- function API - external

It feels like these might be better separated as packages to avoid someone accidentally coupling them together (eg, if one is meant for third parties to build on, and one is meant for internal use, it seems like they should be in separate packages, one for npm publishing, one for bundling w/ the binary).

Could be missing something, a final synchronous code review after the initial comments have been addressed might be good.

Reply:
Yeah I think where we landed is that it's more that the object API construction is internal only, function API is external - the function API contains the ability to call with a `framework`. 

We don't have to publish it independently to start out because the aim is to have most folks using the Object API directly, but when we do, it should be possible to import this & utilize similar to how we're using internally.